### ISSUE TITLE
Option to force a specific protocol

### ISSUE THREAD

**Issue by stefan-guggisberg** [adobe/helix-fetch#8]
**Is your feature request related to a problem? Please describe.**
`helix-fetch` is transparently using the built-in `http`, `https` and `http2` modules und the hood. The majority of testing tools like `nock` or `pollyjs` however only support the `http`/`https` modules. 

**Describe the solution you'd like**
Provide an option to force e.g. the HTTP/1(.1) protocol.

**Additional context**
https://github.com/adobe/helix-pipeline/pull/592#issuecomment-588685408

**Comment by tripodsan** [comment]
another helpful tools would be to have a [PollyJS Adapter](https://netflix.github.io/pollyjs/#/adapters/custom) for `http2`.

**Comment by stefan-guggisberg** [comment]
> another helpful tool would be to have a [PollyJS Adapter](https://netflix.github.io/pollyjs/#/adapters/custom) for `http2`.

the PollyJS Node HTTP Adapter relies on `nock` for intercepting `http(s)` requests. There's a `nock` issue regarding HTTP/2 support but it seems to be stale:  https://github.com/nock/nock/issues/1625

**Comment by tripodsan** [comment]
@stefan-guggisberg any progress on this? it would be good to have this soon....

**Comment by stefan-guggisberg** [comment]
@tripodsan It turned out to be more complicated than expected :/  I'd have to work out a PR to `fetch-h2` and the maintainer doesn't seem to be very responsive. I am also considering alternatives to using `fetch-h2`, see https://github.com/adobe/helix-fetch/issues/14. 

On a different note, how would an option to force a protocol help with the testing problems? When testing a component which is using `fetch-h2` how would you set that option?

**Comment by tripodsan** [comment]
> When testing a component which is using `fetch-h2` how would you set that option?

the context has the `http-protocols` configuration. I'd hoped that setting it to `http1` for force `fetch-h2` to always use the http stack...

https://github.com/grantila/fetch-h2#http-protocols

**Comment by tripodsan** [comment]
ok, I tested it again and it works, when specifying the options in when creating the context (my error was, that I specified them as request options). 

```
const ctx = context({
  userAgent: 'helix-fetch',
  overwriteUserAgent: true,
  httpProtocol: 'http1',
  httpsProtocols: ['http1'],
});
```

So it would be good, we helix-fetch exposes a way to create a custom context.

one caveat is, that it still tries to connect to the host, and not 100% uses the http/https modules. I think it still does the protocol negotiation, and only later applies the restrictions defined in the options. this is sill a problem for offline testing....

**Comment by stefan-guggisberg** [comment]
> So it would be good, we helix-fetch exposes a way to create a custom context.

Agreed, I had that in mind already. I've created an issue:  #15

**Comment by stefan-guggisberg** [comment]
> one caveat is, that it still tries to connect to the host, and not 100% uses the http/https modules. I think it still does the protocol negotiation, and only later applies the restrictions defined in the options. this is sill a problem for offline testing....

yes, that's what I meant by "more complicated than expected". This would need to be addressed in `fetch-h2` and the maintainer is not very responsive.

**Comment by stefan-guggisberg** [comment]
fixed via  #15 and #17
